File: Documentation/v20.1/Content/Samples/Samplescsharp/AcDbMgdExtensions/DatabaseServices/BlockTableExtensionsCommands.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.EditorInput;
using Autodesk.AutoCAD.Runtime;

[assembly: CommandClass(typeof(Samplescsharp.AcDbMgdExtensions.DatabaseServices.BlockTableExtensionsCommands))]
namespace Samplescsharp.AcDbMgdExtensions.DatabaseServices
{
    public class BlockTableExtensionsCommands : CommandClass
    {
        #region GetBlockTableRecords
        [CommandMethod("GetBlockTableRecords")]
        public void GetBlockTableRecords()
        {
            using (Transaction trx = Db.TransactionManager.StartTransaction())
            {
                BlockTable bt = Db.BlockTable();
                var blocks = bt.GetBlockTableRecords();
                foreach (var blk in blocks)
                {
                    Ed.WriteLine($"Contains {blk.GetObjectIds().Count()} this many entities in BlockTableRecord");
                }
            }
        }
        #endregion

        #region GetBlockTableRecordsOpenForWrite
        [CommandMethod("GetBlockTableRecordsOpenForWrite")]
        public void GetBlockTableRecordsOpenForWrite()
        {
            using (Transaction trx = Db.TransactionManager.StartTransaction())
            {
                BlockTable bt = Db.BlockTable();
               var blocks = bt.GetBlockTableRecords(OpenMode.ForWrite);
              

                foreach (var blk in blocks)
                {
                    Ed.WriteLine($"Contains {blk.GetObjectIds().Count()} this many entities in BlockTableRecord");
                }
            }
        }
        #endregion

        #region GetBlockTableRecordsOpenForReadAndXrefBlocks
        [CommandMethod("GetBlockTableRecordsOpenForReadAndXrefBlocks")]
        public void GetBlockTableRecordsOpenForReadAndXrefBlocks()
        {
            using (Transaction trx = Db.TransactionManager.StartTransaction())
            {
                BlockTable bt = Db.BlockTable();
                var blocks = bt.GetBlockTableRecords(OpenMode.ForRead, SymbolTableRecordFilter.IncludeDependent);

                foreach (var blk in blocks)
                {
                    Ed.WriteLine($"Contains {blk.GetObjectIds().Count()} this many entities in BlockTableRecord");
                }
            }
        }
        #endregion
    }
}

File: Documentation/v20.1/Content/Samples/Samplesvb/AcDbMgdExtensions/DatabaseServices/BlockTableExtensionsCommands.vb
Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput
Imports Autodesk.AutoCAD.Runtime

<Assembly: CommandClass(GetType(Samplesvb.AcDbMgdExtensions.DatabaseServices.BlockTableExtensionsCommands))>

Namespace AcDbMgdExtensions.DatabaseServices
    Public Class BlockTableExtensionsCommands
        Inherits CommandClass
        '#Region "GetBlockTableRecords"
        <CommandMethod("GetBlockTableRecords")>
        Public Sub GetBlockTableRecords()
            Using trx As Transaction = Db.TransactionManager.StartTransaction()
                Dim bt As BlockTable = Db.BlockTable()
                Dim blocks = bt.GetBlockTableRecords()
                For Each blk As Object In blocks
                    Ed.WriteLine(String.Format("Contains {0} this many entities in BlockTableRecord", blk.GetObjectIds().Count()))
                Next
            End Using
        End Sub
        '#End Region

        '#Region "GetBlockTableRecordsOpenForWrite"
        <CommandMethod("GetBlockTableRecordsOpenForWrite")>
        Public Sub GetBlockTableRecordsOpenForWrite()
            Using trx As Transaction = Db.TransactionManager.StartTransaction()
                Dim bt As BlockTable = Db.BlockTable()
                Dim blocks = bt.GetBlockTableRecords(OpenMode.ForWrite)


                For Each blk As Object In blocks
                    Ed.WriteLine(String.Format("Contains {0} this many entities in BlockTableRecord", blk.GetObjectIds().Count()))
                Next
            End Using
        End Sub
        '#End Region

        '#Region "GetBlockTableRecordsOpenForReadAndXrefBlocks"
        <CommandMethod("GetBlockTableRecordsOpenForReadAndXrefBlocks")>
        Public Sub GetBlockTableRecordsOpenForReadAndXrefBlocks()
            Using trx As Transaction = Db.TransactionManager.StartTransaction()
                Dim bt As BlockTable = Db.BlockTable()
                Dim blocks = bt.GetBlockTableRecords(OpenMode.ForRead, SymbolTableRecordFilter.IncludeDependent)

                For Each blk As Object In blocks
                    Ed.WriteLine(String.Format("Contains {0} this many entities in BlockTableRecord", blk.GetObjectIds().Count()))
                Next
            End Using
        End Sub
        '#End Region
    End Class
End Namespace

